### PR TITLE
Allow transaction receipt in balance matchers

### PR DESCRIPTION
**Pull request by duckception**

### CODE REVIEW

File: waffle-chai/src/matchers/changeBalances.ts [hunk 1]
```diff
@@ -38,17 +34,31 @@ export function supportChangeBalances(Assertion: Chai.AssertionStatic) {
 }
 
 async function getBalanceChanges(
-  transactionCallback: () => any,
+  transaction: providers.TransactionResponse | Function,
```
Comment (SpidiRitsu) [inline waffle-chai/src/matchers/changeBalances.ts:37]:
```ts
transaction: providers.TransactionResponse | () => Promise<providers.TransactionResponse>,
```

Comment (SpidiRitsu) [inline waffle-chai/src/matchers/changeBalances.ts:37]: Dont use Function, use `() => Promise<void> | void`

File: waffle-chai/src/matchers/changeBalances.ts [hunk 2]
```diff
@@ -38,17 +34,31 @@ export function supportChangeBalances(Assertion: Chai.AssertionStatic) {
...
 ) {
-  const balancesBefore = await Promise.all(
-    signers.map(getBalanceOf)
-  );
-  await transactionCallback();
-  const balancesAfter = await Promise.all(
-    signers.map(getBalanceOf)
-  );
-  return balancesAfter.map((balance, ind) => balance.sub(balancesBefore[ind]));
+  if (transaction instanceof Function) {
```
Comment (SpidiRitsu) [inline waffle-chai/src/matchers/changeBalances.ts:40]: `typeof transaction === 'function'`

File: waffle-chai/src/matchers/misc/account.ts [hunk 3]
```diff
@@ -1,12 +1,19 @@
 import {Contract, Signer} from 'ethers';
+import {ensure} from '../calledOnContract/utils';
 
 export type Account = Signer | Contract;
 
-export async function getBalanceOf(account: Account) {
-  if (account instanceof Contract) {
-    return account.provider.getBalance(account.address);
+export async function getBalanceOf(account: Account, blockNumber?: number) {
+  if (blockNumber) {
```
Comment (SpidiRitsu) [inline waffle-chai/src/matchers/misc/account.ts:7]: `account.provider.getBalance(getAddressOf(account), blockNumber)`

File: waffle-chai/src/matchers/changeBalances.ts [hunk 4]
```diff
@@ -8,10 +8,6 @@ export function supportChangeBalances(Assertion: Chai.AssertionStatic) {
     balanceChanges: any[]
   ) {
     const subject = this._obj;
-    if (typeof subject !== 'function') {
-      throw new Error(`Expect subject should be a callback returning the Promise
-        e.g.: await expect(() => wallet.send({to: '0xb', value: 200})).to.changeBalances(['0xa', '0xb'], [-200, 200])`);
```
Comment (SpidiRitsu) [inline waffle-chai/src/matchers/changeBalances.ts:13]: changeBalances({ '0xa': -200, '0xb': 200 })

File: waffle-chai/src/matchers/misc/account.ts [hunk 5]
```diff
@@ -2,14 +2,6 @@ import {Contract, Signer} from 'ethers';
 
 export type Account = Signer | Contract;
 
-export async function getBalanceOf(account: Account) {
-  if (account instanceof Contract) {
-    return account.provider.getBalance(account.address);
-  } else {
-    return account.getBalance();
-  }
-}
```
Comment (marekkirejczyk) [inline waffle-chai/src/matchers/misc/account.ts:11]: 👍

**Review by marekkirejczyk — APPROVED** [review]

**Pull request merged**